Pydist overrides

anshprat opened the pull request.

Commit:
Adding pydist-overrides support via config
Supports yaml in config file
example:
pydist_override:
  python-websocket: python-websocket-client

	modified:   aasemble/django/apps/buildsvc/pkgbuild/python.py

Commit:
Fixing tox/travis errors

anshprat commented:
Please review. Would love some help with adding test cases for this.

pljensen commented on aasemble/django/apps/buildsvc/pkgbuild/python.py:
```diff
@@ -41,7 +43,34 @@ def binary_pkg_name(self):
+        if len(override_dict) == 0:
+            self.logger.info('No pydist-overrides in config')
+            return
+        pydist_overrides_str = (yaml.dump(override_dict, default_flow_style=False)).replace(':', '')
```
Use a Django template instead, so that we're actually in control of the output.

anshprat replied:
I thought of doing so first - but was not sure of how to do many to one mapping. Basically the template would ve one place holder variable - but the config file can have many entries. So should the substitution should be the many entries in config as a single string to the template variable?

pljensen replied:
Django templates let you iterate:

```
{% for elem in somelist %}
{{ elem.whatever }}
{% endfor %}
```

pljensen commented on aasemble/django/apps/buildsvc/pkgbuild/python.py:
```diff
@@ -41,7 +43,34 @@ def binary_pkg_name(self):
+
+        if os.path.exists(pydist_overrides):
+            with open(pydist_overrides, 'r') as fp:
+                pre_pydist_overrides = fp.read()
```
If there's a debian/ dir, how did we end up in the python handler?

anshprat replied:
Presently this is not the case - but will be adding support for debian folder + python builder capability.

pljensen commented on aasemble/django/apps/buildsvc/pkgbuild/python.py:
```diff
@@ -41,7 +43,34 @@ def binary_pkg_name(self):
+            fp.write(pydist_overrides_str)
+            fp.write(pre_pydist_overrides)
+            self.logger.info('Saving pydist-overrides file: %s' % (pydist_overrides))
+            return
```
Does the order matter? If not, don't bother reading the existing pydist-overrides first, just open the file in append mode (mode 'a' instead of 'w') and add the new stuff.

anshprat replied:
ok..

pljensen commented on aasemble/django/apps/buildsvc/pkgbuild/python.py:
```diff
@@ -41,7 +43,34 @@ def binary_pkg_name(self):
+            return
+
     def detect_runtime_dependencies(self):
+        super(PythonBuilder, self).populate_debian_dir()
```
This seems strange. Why not override populate_debian_dir since that's what you're calling with super()?

anshprat replied:
Not sure about this completely - can you give an example way of doing this without interfering with other builders?

pljensen replied:
Well, instead of overriding detect_runtime_dependencies, you just override populate_debian_dir. It won't interfere with other builders. Only ones that inherit from this builder (which none do).

Basically, just remove your changes to detect_runtime_dependencies and instead add:

```
def populate_debian_dir(self):
    super(PythonBuilder, self).populate_debian_dir()
    self.add_pydist_overrides()
```

anshprat replied:
ok..

Commit:
Changes as per review

	modified:   aasemble/django/apps/buildsvc/pkgbuild/python.py
	new file:   aasemble/django/apps/buildsvc/pkgbuild/templates/pkgbuild/pydist-overrides.tmpl

Commit:
pep8/travis/tox/flake fixes

pljensen commented:
This looks good now!

Can you provide an example of how to use this? Add it to the docs or include it in the commit message so I can create a some documentation for it later.

anshprat commented:
Example yaml - https://github.com/anshprat/docker-py/commit/5bb349cf3f019f646353bfc789a47ee1ccf0891d
